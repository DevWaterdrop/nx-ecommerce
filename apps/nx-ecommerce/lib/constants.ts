export const LOVELY_ERROR = 'I HATE TYPESCRIPT+STRAPI+GRAPHQL';
export const MAXIMUM_IN_CART = 16;

export enum PRICE_FILTER {
  ZERO_TWENTY = 'zt',
  TWENTY_FORTY = 'tf',
  FORTY_EIGHTY = 'fe',
  EIGHTY_PLUS = 'ep',
}

export enum SORT_FILTER {
  PRICE_LOW_HIGH = 'plh',
  PRICE_HIGH_LOW = 'phl',
  NAME = 'name',
}

export const PRICES = [
  { value: PRICE_FILTER.ZERO_TWENTY, label: '0.00$ - 19.99$' },
  { value: PRICE_FILTER.TWENTY_FORTY, label: '20.00$ - 39.99$' },
  { value: PRICE_FILTER.FORTY_EIGHTY, label: '40.00$ - 79.99$' },
  { value: PRICE_FILTER.EIGHTY_PLUS, label: '80.00$+' },
];

export const SORT = [
  { value: SORT_FILTER.PRICE_LOW_HIGH, label: 'Price: low to high' },
  { value: SORT_FILTER.PRICE_HIGH_LOW, label: 'Price: high to low' },
  { value: SORT_FILTER.NAME, label: 'Name' },
];
