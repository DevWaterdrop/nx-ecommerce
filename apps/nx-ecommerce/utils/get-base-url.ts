export function getBaseURL(path = '') {
  return `${
    (process.env['NX_BASE_URL'] as string) || 'http://localhost:1337'
  }${path}`;
}
