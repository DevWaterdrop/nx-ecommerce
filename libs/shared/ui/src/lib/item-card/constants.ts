import { ItemCardProps } from './item-card';
import flowerJPEG from '../../props/flower.jpeg';
import roomJPEG from '../../props/room.jpeg';

const ITEM: Required<ItemCardProps['item']> = {
  price: 65,
  name: 'Lovely flower',
  id: '#',
  images: [flowerJPEG, roomJPEG],
  smallDescription: 'For you, for me, for us',
  tag: 'new',
};

export const DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleClick: () => console.log('clicked'),
};
