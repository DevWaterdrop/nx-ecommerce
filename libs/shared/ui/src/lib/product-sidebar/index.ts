import { ProductSidebar, ProductSidebarProps } from './product-sidebar';

export { ProductSidebar };
export type { ProductSidebarProps };
