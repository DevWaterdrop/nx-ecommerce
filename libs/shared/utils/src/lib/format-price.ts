export function formatPrice(price: number) {
  return `${Math.max(price, 0).toFixed(2)}$`;
}
