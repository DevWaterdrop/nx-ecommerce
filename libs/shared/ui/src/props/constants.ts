import { ButtonProps } from '../lib/button';
import { CartItemProps } from '../lib/cart-item';
import { CartSidebarProps } from '../lib/cart-sidebar';
import { CategoryCardProps } from '../lib/category-card';
import { IconProps } from '../lib/icon/icon';
import { ItemCardProps } from '../lib/item-card';
import { NavButtonProps } from '../lib/nav-button';
import { OrdersTableProps } from '../lib/orders-table';
import { ProductGalleryProps } from '../lib/product-gallery';
import { ProductInformationProps } from '../lib/product-information';
import { ProductSidebarProps } from '../lib/product-sidebar';
import { SearchbarProps } from '../lib/searchbar';
import { SearchbarItemProps } from '../lib/searchbar-item';
import { SelectProps } from '../lib/select';
import flowerJPEG from './flower.jpeg';
import roomJPEG from './room.jpeg';

export const ITEM: ItemCardProps['item'] = {
  slug: 'lovely-flower',
  price: 65,
  name: 'Lovely flower',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'For you, for me, for us',
  description: 'Lovely description',
};

export const ITEM_SECOND: ItemCardProps['item'] = {
  slug: 'roomy',
  price: 50,
  name: 'Roomy',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'Hmmm, nice vase',
  description: 'Vaaaase',
};

export const ITEM_CARD_DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleCartClick: () => console.log('cart clicked'),
  handleAmountCartClick: () => console.log('amount cart clicked'),
  handleFavoriteClick: () => console.log('favorite clicked'),
};

export const CATEGORY_CARD_DEFAULT_PROPS: CategoryCardProps = {
  category: {
    smallDescription:
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
    image: {
      data: {
        attributes: {
          url: roomJPEG,
        },
      },
    },
    slug: '1',
  },
};

export const NAV_BUTTON_DEFAULT_PROPS: NavButtonProps = {
  type: 'heart',
};

export const ICON_DEFAULT_PROPS: IconProps = {
  icon: 'cart',
};

export const BUTTON_DEFAULT_PROPS: ButtonProps = {
  children: 'hello',
  tag: 'button',
  size: 'base',
  type: 'primary',
  round: 'full',
};

export const SEARCHBAR_ITEM: SearchbarProps['items'][number] = {
  id: '111',
  attributes: {
    slug: 'lovely-flower',
    smallDescription: 'For you, for me, for us',
    name: 'Lovely flower',
    images: {
      data: [
        {
          attributes: {
            url: flowerJPEG,
          },
        },
      ],
    },
  },
};

export const SEARCHBAR_ITEM_DEFAULT_PROPS: SearchbarItemProps = {
  item: {
    slug: 'lovely-flower',
    smallDescription: 'For you, for me, for us',
    name: 'Lovely flower',
    images: {
      data: [
        {
          attributes: {
            url: flowerJPEG,
          },
        },
      ],
    },
  },
  handleClick: () => void 1,
};

export const SELECT_ITEM = {
  label: 'Label1',
  value: 'value1',
};

export const SELECT_DEFAULT_PROPS: Required<
  Omit<SelectProps, 'placeholder' | 'multiple'>
> = {
  items: [SELECT_ITEM],
  initial: ['value1'],
  onChange: () => console.log('changed'),
};

export const PRODUCT: ProductSidebarProps['product'] = {
  slug: 'lovely-flower',
  price: 65,
  name: 'Lovely flower',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'For you, for me, for us',
  description: 'Lovely description',
  information: [
    {
      title: 'Title1',
      content: 'content1',
    },
  ],
};

export const PRODUCT_SIDEBAR_DEFAULT_PROPS: ProductSidebarProps = {
  product: PRODUCT,
  handleCartClick: () => {
    console.log('cart');
  },
  handleRemoveCartClick: () => {
    console.log('remove cart');
  },
  handleFavoriteClick: () => {
    console.log('favorite');
  },
  cartValue: new Map([]),
  favoritesValue: new Set('lovely-flower'),
};

export const PRODUCT_INFORMATION_DEFAULT_PROPS: ProductInformationProps = {
  information: [...Array(10).keys()].map((_value, index) => ({
    title: `Title${index + 1}`,
    content: `content${index + 1}`,
  })),
};

export const PRODUCT_GALLERY_DEFAULT_PROPS: ProductGalleryProps = {
  images: [
    {
      attributes: {
        url: flowerJPEG,
      },
    },
    {
      attributes: {
        url: roomJPEG,
      },
    },
  ],
};

const PRODUCT_ORDER_TABLE: OrdersTableProps['orders'][number]['items'][number]['product'] =
  {
    name: 'Flower',
    slug: 'slug-flower',
    description: '',
    smallDescription: '',
    price: 50,
  };

export const ORDERS_TABLE_DEFAULT_PROPS: OrdersTableProps = {
  orders: [
    {
      id: 'id-1',
      amount: 50,
      items: [{ id: '1', amount: 1, product: PRODUCT_ORDER_TABLE }],
    },
    {
      id: 'id-2',
      amount: 100,
      items: [{ id: '2', amount: 2, product: PRODUCT_ORDER_TABLE }],
    },
  ],
};

export const CART_SIDEBAR_DEFAULT_PROPS: CartSidebarProps = {
  cartTotal: 100,
  isUser: true,
  handleCheckoutClick: async () => {
    console.log('checkout');
  },
  isLoading: false,
};

export const CART_ITEM_DEFAULT_PROPS: CartItemProps = {
  inCartAmount: 1,
  item: PRODUCT,
  handleCartMinusClick: () => {
    console.log('cart minus');
  },
  handleCartPlusClick: () => {
    console.log('cart plus');
  },
  handleRemoveFromCart: () => {
    console.log('cart remove');
  },
  toggleFavorite: () => {
    console.log('favorite');
  },
};
