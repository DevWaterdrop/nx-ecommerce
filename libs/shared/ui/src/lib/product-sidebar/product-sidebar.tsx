import { ProductSlugQuery } from '@nx-ecommerce/shared/graphql/types';
import { formatPrice } from '@nx-ecommerce/shared/utils/format-price';
import clsx from 'clsx';
import { Cart } from './local-components/cart';
import { Favorite } from './local-components/favorite';

export interface ProductSidebarProps {
  product: NonNullable<
    NonNullable<ProductSlugQuery['products']>['data'][0]['attributes']
  >;
  handleCartClick: () => void;
  handleRemoveCartClick: () => void;
  handleFavoriteClick: () => void;
  cartValue: Map<string, number>;
  favoritesValue: Set<string>;
}

export const ProductSidebar: React.FC<ProductSidebarProps> = (props) => {
  const {
    product,
    handleCartClick,
    handleFavoriteClick,
    handleRemoveCartClick,
    cartValue,
    favoritesValue,
  } = props;

  const { name, price, smallDescription, slug } = product;

  return (
    <aside
      className={clsx(
        'h-full min-w-56 w-full',
        'flex flex-col',
        'md:(order-2 top-12 sticky w-2/5)'
      )}
    >
      <div className={clsx('mb-2', 'flex items-center justify-between')}>
        <h1
          className={clsx(
            'max-w-3/5',
            'font-bold text-3xl uppercase break-words'
          )}
        >
          {name}
        </h1>
        <span className="font-semibold text-base">{formatPrice(price)}</span>
      </div>
      <p className={clsx('text-sm', 'opacity-75')}>{smallDescription}</p>
      <div
        className={clsx(
          'mt-8 py-6 px-4',
          'flex flex-col gap-1',
          'border rounded-md',
          'text-base',
          'md:(text-sm py-4)'
        )}
      >
        <p>100% of the net profits will be donated to Ukraine!</p>
      </div>
      <div className={clsx('mt-8', 'flex gap-4 justify-between')}>
        <Cart
          handleCartClick={handleCartClick}
          cartValue={cartValue}
          handleRemoveCartClick={handleRemoveCartClick}
          slug={slug}
        />
        <Favorite
          handleFavoriteClick={handleFavoriteClick}
          slug={slug}
          favoritesValue={favoritesValue}
        />
      </div>
    </aside>
  );
};
