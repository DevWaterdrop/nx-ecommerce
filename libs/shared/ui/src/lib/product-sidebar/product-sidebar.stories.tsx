import { Story, Meta } from '@storybook/react';
import { PRODUCT_SIDEBAR_DEFAULT_PROPS } from '../../props/constants';
import { ProductSidebar, ProductSidebarProps } from './product-sidebar';

export default {
  component: ProductSidebar,
  title: 'Components/Product/Sidebar',
} as Meta;

const Template: Story<ProductSidebarProps> = (args) => (
  <ProductSidebar {...args} />
);

export const Primary = Template.bind({});
Primary.args = {
  ...PRODUCT_SIDEBAR_DEFAULT_PROPS,
};
