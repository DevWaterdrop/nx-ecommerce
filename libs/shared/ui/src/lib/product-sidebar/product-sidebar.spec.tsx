import { render, RenderResult, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { ProductSidebar } from './product-sidebar';
import { PRODUCT_SIDEBAR_DEFAULT_PROPS } from '../../props/constants';
import { formatPrice } from '@nx-ecommerce/shared/utils/format-price';

describe('ProductSidebar', () => {
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(
      <ProductSidebar {...PRODUCT_SIDEBAR_DEFAULT_PROPS} />
    );
  });

  it('should render successfully', () => {
    const { baseElement } = renderResult;
    expect(baseElement).toBeTruthy();
  });

  it('should have proper information', () => {
    const { smallDescription, name, price } =
      PRODUCT_SIDEBAR_DEFAULT_PROPS.product;

    expect(screen.getByText(smallDescription)).toBeInTheDocument();

    expect(screen.getByText(name)).toBeInTheDocument();

    expect(screen.queryByText(price)).not.toBeInTheDocument();
    expect(screen.queryByText(formatPrice(price))).toBeInTheDocument();

    const bannerInfo = '100% of the net profits will be donated to Ukraine!';
    expect(screen.queryByText(bannerInfo)).toBeInTheDocument();
  });
});
