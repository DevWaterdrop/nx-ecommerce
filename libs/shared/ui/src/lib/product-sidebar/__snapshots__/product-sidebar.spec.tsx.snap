// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductSidebar should render successfully 1`] = `
<aside
  className="h-full min-w-56 w-full flex flex-col md:(order-2 top-12 sticky w-2/5)"
>
  <div
    className="mb-2 flex items-center justify-between"
  >
    <h1
      className="max-w-3/5 font-bold text-3xl uppercase break-words"
    >
      Lovely flower
    </h1>
    <span
      className="font-semibold text-base"
    >
      65.00$
    </span>
  </div>
  <p
    className="text-sm opacity-75"
  >
    For you, for me, for us
  </p>
  <div
    className="mt-8 py-6 px-4 flex flex-col gap-1 border rounded-md text-base md:(text-sm py-4)"
  >
    <p>
      100% of the net profits will be donated to Ukraine!
    </p>
  </div>
  <div
    className="mt-8 flex gap-4 justify-between"
  >
    <div
      className="flex-1 relative"
    >
      <button
        className="inline-block cursor-pointer transform-gpu transition-all active:scale-92 rounded-full w-full h-full text-white bg-sky-600 hover:(bg-sky-700) active:(bg-sky-800)"
        onClick={[Function]}
      >
        <span
          className="font-semibold"
        >
          Add to bag
        </span>
      </button>
    </div>
    <button
      className="inline-block cursor-pointer transform-gpu transition-all active:scale-92 rounded-full p-4 hover:(bg-gray-100 border-black) active:(bg-gray-300) border border-gray-300 flex justify-center items-center"
      onClick={[Function]}
      title="add to favorite"
      type="button"
    >
      <div
        aria-hidden="true"
        className="relative flex items-center select-none pointer-events-none h-4 w-4  sm:(h-6 w-6)"
      >
        Next image stub
      </div>
    </button>
  </div>
</aside>
`;
